src/extra/recipes/grating.py

```diff
@@ -109,10 +153,39 @@ def _fromdict(self, all_data):
         for k, v in all_data.items():
             setattr(self, k, v)
 
+    def guess_offset(self):
+        """
+        Guess offset.
+        """
+        I = self._grating_signal.xarray().sel(dim_1=np.s_[self.min_pixel:self.max_pixel]).mean('dim_1').mean('trainId')
+        threshold = np.median(I)
+        self.offset = np.where(I > threshold)[0][0]
+        logging.info(f"Offset estimated at {self.offset}")
+
+    def get_pulse_period(self, pulses: XrayPulses):
+        """
+        Estimate difference between samples in neighbour pulses.
+
+        Args:
+          pulses: XrayPulses element for the run.
+
+        Returns: Sample difference between neighbour pulses.
+        """
+        pulse_ids = pulses.pulse_ids(labelled=True)
+        pids_by_train = pulse_ids.groupby(level=0)
+        pulse_period = int(pids_by_train.diff().min())
```

Comment:
Out of curiosity, do I understand correctly that this is computing the period between pulses in units of entries in the bunch pattern table (or global pulse ID)?